source/internal-communication.rst

```diff
@@ -0,0 +1,539 @@
+
+#. JWT generation. The documentation for this plugin is available at `Kong JWT
+   plugin page`_.
+#. Configuration a plugin which will forward all policies requests to Auth.
+   will invoke Auth in order to authenticate requests. This plugin is available
+   in `PEP-Kong repository`_.
+
+
+Emitted messages
+****************
+
+Auth will emit just one message via Kafka for tenant creation:
+
+.. code-block:: json
+
+   {
+     "type" : "CREATE",
+     "tenant" : "XYZ"
+   }
+
+Device Manager
+--------------
+
+DeviceManager stores and retrieves information models for devices and templates
+and a few static information about them as well. Whenever a device is created,
+removed or just edited, it will publish a message through Kafka. It depends
+only on DataBroker and Kafka for reasons already explained in this document.
+
+IoT agent
+---------
+
+IoT agents receive messages from devices and translate them into a default
+message to be published to other components. In order to do that, they might
+want to know which devices are created in order to properly filter messages
+which are not allowed into dojot (using, for instance, security information to
+block messages from unauthorized devices). It will use the ``device-data``
+subject and bootstrap tenants as described in `Bootstrapping tenants`_.
+
+After requesting the topics for all tenants within `device-data` subject, IoT
+agent will start receiving data from devices. As there are a plethora of ways
+by which devices can do that, this step won't be detailed in this section (this
+is highly dependent on how each IoT agent works). It must, though, send a
+message to Kafka to inform other components of all new data that the device
+just sent. This is shown in :numref:`IoT agent - kafka`.
+
+.. _IoT agent - kafka:
+.. uml::
+   :caption: IoT agent message to Kafka
+   :align: center
+
+   control Kafka
+
+   IoTAgent -> Kafka: SEND MESSAGE\n topic:890874987ef...\ndata: IoTAgentMessage
+   Kafka -> IoTAgent: OK
+
+
+The data sent by IoT agent has the structure shown in :numref:`IoT agent
+message`.
+
+.. _IoT agent message:
+.. uml::
+   :caption: IoT agent message structure
+   :align: center
+
+
+   class Metadata {
+     + deviceid: string
+     + tenant: string
+     + timestamp: long int
+    }
+
+    class IoTAgentMessage {
+      + metadata: Metadata
+      + attrs: Dict<string, any>
+    }
+
+    IoTAgentMessage::metadata -> Metadata
+
+Such message would be:
+
+.. code-block:: json
+
+    {
+        "metadata": {
+            "deviceid": "c6ea4b",
+            "tenant": "admin",
+            "timestamp": 1528226137452
+        },
+        "attrs": {
+            "humidity": 60,
+            "temperature" : 23
+        }
+    }
+
+
+
+Persister
+---------
+
+Persister is a very simple service which only purpose is to receive messages
+from devices (using ``device-data`` subject) and store them into MongoDB. For
+that, the bootstrapping procedure (detailed in `Bootstrapping tenants`_) is
```

Comment:
Add a reference to the bootstraping tenant section

Reply:
Didn't get it. Isn't that `Bootstrapping tenants_`?